test/e2e/native_histograms_test.go

```diff
@@ -0,0 +1,182 @@
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/prometheus/prometheus/storage/remote"
+	"github.com/thanos-io/thanos/pkg/promclient"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"github.com/thanos-io/thanos/pkg/testutil"
+	"github.com/thanos-io/thanos/test/e2e/e2ethanos"
+)
+
+func TestQueryNativeHistograms(t *testing.T) {
+	e, err := e2e.NewDockerEnvironment("nat-hist-query")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	prom, sidecar := e2ethanos.NewPrometheusWithSidecar(e, "prom", e2ethanos.DefaultPromConfig("prom-alone", 0, "", "", e2ethanos.LocalPrometheusTarget), "", "quay.io/prometheus/prometheus:v2.40.5", "", "native-histograms", "remote-write-receiver")
+	testutil.Ok(t, e2e.StartAndWaitReady(prom, sidecar))
+
+	querier := e2ethanos.NewQuerierBuilder(e, "querier", sidecar.InternalEndpoint("grpc")).Init()
+	testutil.Ok(t, e2e.StartAndWaitReady(querier))
+
+	rawRemoteWriteURL := "http://" + prom.Endpoint("http") + "/api/v1/write"
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	t.Cleanup(cancel)
+
+	ts := time.Now()
+
+	getTs := func() time.Time {
+		return ts
+	}
+
+	writeRequest(t, ctx, rawRemoteWriteURL, nativeHistogramWriteRequest(ts))
+
+	// Make sure we can query native histogram directly from Prometheus.
+	queryAndAssertSeries(t, ctx, prom.Endpoint("http"), func() string { return "test_histogram" }, getTs, promclient.QueryOptions{}, []model.Metric{
+		{
+			"__name__": "test_histogram",
+			"foo":      "bar",
+		},
+	})
+
+	queryAndAssertError(t, ctx, querier.Endpoint("http"), func() string { return "test_histogram" }, getTs, promclient.QueryOptions{}, "invalid chunk encoding")
+}
+
+func TestWriteNativeHistograms(t *testing.T) {
+	e, err := e2e.NewDockerEnvironment("nat-hist-write")
+	testutil.Ok(t, err)
+	t.Cleanup(e2ethanos.CleanScenario(t, e))
+
+	receiver := e2ethanos.NewReceiveBuilder(e, "receiver").WithIngestionEnabled().Init()
+	testutil.Ok(t, e2e.StartAndWaitReady(receiver))
+
+	querier := e2ethanos.NewQuerierBuilder(e, "querier", receiver.InternalEndpoint("grpc")).Init()
+	testutil.Ok(t, e2e.StartAndWaitReady(querier))
+
+	rawRemoteWriteURL := "http://" + receiver.Endpoint("remote-write") + "/api/v1/receive"
+
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	t.Cleanup(cancel)
+
+	writeRequest(t, ctx, rawRemoteWriteURL, nativeHistogramWriteRequest(time.Now()))
+
+	queryAndAssertSeries(t, ctx, querier.Endpoint("http"), func() string { return "test_histogram" }, time.Now, promclient.QueryOptions{}, []model.Metric{})
+	queryAndAssertSeries(t, ctx, querier.Endpoint("http"), func() string { return "test_sample" }, time.Now, promclient.QueryOptions{}, []model.Metric{
+		{
+			"__name__":  "test_sample",
+			"bar":       "foo",
+			"receive":   "receive-receiver",
+			"tenant_id": "default-tenant",
+		},
+	})
+}
+
+func writeRequest(t *testing.T, ctx context.Context, rawRemoteWriteURL string, req *prompb.WriteRequest) {
```

Comment:
Maybe the end of `synthesizeSamples` function could reuse this one to avoid duplication? In there, we create a `*prompb.WriteRequest` and send it off just like in this function.